File: rewrite-java/src/main/java/org/openrewrite/java/JavaTypeVisitor.java
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java;

import lombok.Getter;
import lombok.Setter;
import org.jspecify.annotations.Nullable;
import org.openrewrite.Cursor;
import org.openrewrite.internal.ListUtils;
import org.openrewrite.java.tree.JavaType;

import java.util.IdentityHashMap;
import java.util.List;
import java.util.Set;

import static java.util.Collections.newSetFromMap;
import static org.openrewrite.internal.ListUtils.arrayOrNullIfEmpty;

@Getter
@Setter
public class JavaTypeVisitor<P> {
    private Cursor cursor = new Cursor(null, "root");

    /**
     * Track types that have been visited to prevent infinite recursion and exponential
     * explosion from revisiting the same types via different paths.
     * Uses identity comparison since we care about the same object instance.
     */
    @Nullable
    private Set<JavaType> visited;

    public @Nullable <JT extends JavaType> List<JT> visit(@Nullable List<JT> javaTypes, P p) {
        //noinspection unchecked
        return ListUtils.map(javaTypes, jt -> (JT) visit(jt, p));
    }

    @SuppressWarnings("unused")
    public @Nullable JavaType preVisit(JavaType javaType, P p) {
        return javaType;
    }

    @SuppressWarnings("unused")
    public @Nullable JavaType postVisit(JavaType javaType, P p) {
        return javaType;
    }

    /**
     * By calling this method, you are asserting that you know that the outcome will be non-null
     * when the compiler couldn't otherwise prove this to be the case. This method is a shortcut
     * for having to assert the non-nullability of the returned tree.
     *
     * @param javaType A non-null type.
     * @param p        A state object that passes through the visitor.
     * @return A non-null type.
     */
    public JavaType visitNonNull(JavaType javaType, P p) {
        JavaType t = visit(javaType, p);
        //noinspection ConstantConditions
        assert t != null;
        return t;
    }

    public JavaType visit(@Nullable JavaType javaType, P p) {
        if (javaType != null) {
            // Check if we've already visited this type - if so, return it unchanged.
            // This prevents both infinite recursion (cycles) and exponential explosion
            // (revisiting the same types via different paths in the type graph).
            if (visited != null && visited.contains(javaType)) {
                return javaType;
            }

            // Mark this type as visited
            if (visited == null) {
                visited = newSetFromMap(new IdentityHashMap<>());
            }
            visited.add(javaType);

            cursor = new Cursor(cursor, javaType);
            javaType = preVisit(javaType, p);

            // If preVisit returned a different object, also track that to prevent cycles
            if (javaType != null && !visited.contains(javaType)) {
                visited.add(javaType);
            }

            if (javaType instanceof JavaType.Array) {
                javaType = visitArray((JavaType.Array) javaType, p);
            } else if (javaType instanceof JavaType.Annotation) {
                javaType = visitAnnotation((JavaType.Annotation) javaType, p);
            } else if (javaType instanceof JavaType.Class) {
                javaType = visitClass((JavaType.Class) javaType, p);
            } else if (javaType instanceof JavaType.GenericTypeVariable) {
                javaType = visitGenericTypeVariable((JavaType.GenericTypeVariable) javaType, p);
            } else if (javaType instanceof JavaType.Intersection) {
                javaType = visitIntersection((JavaType.Intersection) javaType, p);
            } else if (javaType instanceof JavaType.MultiCatch) {
                javaType = visitMultiCatch((JavaType.MultiCatch) javaType, p);
            } else if (javaType instanceof JavaType.Parameterized) {
                javaType = visitParameterized((JavaType.Parameterized) javaType, p);
            } else if (javaType instanceof JavaType.Primitive) {
                javaType = visitPrimitive((JavaType.Primitive) javaType, p);
            } else if (javaType instanceof JavaType.Method) {
                javaType = visitMethod((JavaType.Method) javaType, p);
            } else if (javaType instanceof JavaType.Variable) {
                javaType = visitVariable((JavaType.Variable) javaType, p);
            } else if (javaType instanceof JavaType.Unknown) {
                javaType = visitUnknown((JavaType.Unknown) javaType, p);
            }

            if (javaType != null) {
                javaType = postVisit(javaType, p);
            }

            cursor = cursor.getParentOrThrow();

            //noinspection ConstantConditions
            return javaType;
        }

        //noinspection ConstantConditions
        return null;
    }

    public JavaType visitMultiCatch(JavaType.MultiCatch multiCatch, P p) {
        return multiCatch.unsafeSet(
                visit(multiCatch.getThrowableTypes(), p)
        );
    }

    public JavaType visitAnnotation(JavaType.Annotation annotation, P p) {
        return annotation.unsafeSet(
                (JavaType.FullyQualified) visit(annotation.getType(), p),
                arrayOrNullIfEmpty(annotation.getValues(), JavaType.EMPTY_ANNOTATION_VALUE_ARRAY)
        );
    }

    public JavaType visitArray(JavaType.Array array, P p) {
        return array.unsafeSet(
                visit(array.getElemType(), p),
                arrayOrNullIfEmpty(visit(array.getAnnotations(), p), JavaType.EMPTY_FULLY_QUALIFIED_ARRAY)
        );
    }

    public JavaType visitClass(JavaType.Class aClass, P p) {
        return aClass.unsafeSet(
                visit(aClass.getTypeParameters(), p),
                (JavaType.FullyQualified) visit(aClass.getSupertype(), p),
                (JavaType.FullyQualified) visit(aClass.getOwningClass(), p),
                ListUtils.map(aClass.getAnnotations(), a -> (JavaType.FullyQualified) visit(a, p)),
                ListUtils.map(aClass.getInterfaces(), i -> (JavaType.FullyQualified) visit(i, p)),
                ListUtils.map(aClass.getMembers(), m -> (JavaType.Variable) visit(m, p)),
                ListUtils.map(aClass.getMethods(), m -> (JavaType.Method) visit(m, p))
        );
    }

    public JavaType visitGenericTypeVariable(JavaType.GenericTypeVariable generic, P p) {
        return generic.unsafeSet(
                generic.getName(),
                generic.getVariance(),
                visit(generic.getBounds(), p)
        );
    }

    public JavaType visitIntersection(JavaType.Intersection intersection, P p) {
        return intersection.unsafeSet(
                visit(intersection.getBounds(), p)
        );
    }

    /**
     * This does not visit the declaring type to avoid a visitor cycle.
     *
     * @param method The method to visit
     * @param p      Visit context
     * @return A method
     */
    public JavaType visitMethod(JavaType.Method method, P p) {
        return method.unsafeSet(
                (JavaType.FullyQualified) visit(method.getDeclaringType(), p),
                visit(method.getReturnType(), p),
                visit(method.getParameterTypes(), p),
                visit(method.getThrownExceptions(), p),
                ListUtils.map(method.getAnnotations(), a -> (JavaType.FullyQualified) visit(a, p))
        );
    }

    public JavaType visitParameterized(JavaType.Parameterized parameterized, P p) {
        return parameterized.unsafeSet(
                (JavaType.FullyQualified) visit(parameterized.getType(), p),
                visit(parameterized.getTypeParameters(), p)
        );
    }

    public JavaType visitPrimitive(JavaType.Primitive primitive, P p) {
        return primitive;
    }

    public JavaType visitUnknown(JavaType.Unknown unknown, P p) {
        return unknown;
    }

    /**
     * This does not visit the owner to avoid a visitor cycle.
     *
     * @param variable The variable to visit
     * @param p        Visit context
     * @return A variable
     */
    public JavaType visitVariable(JavaType.Variable variable, P p) {
        return variable.unsafeSet(
                visit(variable.getOwner(), p),
                visit(variable.getType(), p),
                ListUtils.map(variable.getAnnotations(), a -> (JavaType.FullyQualified) visit(a, p))
        );
    }
}

File: rewrite-go/pkg/visitor/java_type_visitor.go
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package visitor

import "github.com/openrewrite/rewrite/rewrite-go/pkg/tree/java"

// JavaTypeVisitor traverses and optionally transforms JavaType instances.
// Mirrors org.openrewrite.java.JavaTypeVisitor in Java.
// Embed in a struct and override visit methods to customize behavior.
type JavaTypeVisitor struct {
	// Self must point to the outermost embedding struct for virtual dispatch.
	Self JavaTypeVisitorI
}

// JavaTypeVisitorI defines all overridable visit methods for JavaType.
type JavaTypeVisitorI interface {
	Visit(javaType java.JavaType, p any) java.JavaType
	VisitAnnotation(annotation *java.JavaTypeAnnotation, p any) java.JavaType
	VisitArray(array *java.JavaTypeArray, p any) java.JavaType
	VisitClass(class_ *java.JavaTypeClass, p any) java.JavaType
	VisitShallowClass(shallow *java.JavaTypeShallowClass, p any) java.JavaType
	VisitGenericTypeVariable(generic *java.JavaTypeGenericTypeVariable, p any) java.JavaType
	VisitIntersection(intersection *java.JavaTypeIntersection, p any) java.JavaType
	VisitMethod(method *java.JavaTypeMethod, p any) java.JavaType
	VisitMultiCatch(multiCatch *java.JavaTypeMultiCatch, p any) java.JavaType
	VisitParameterized(parameterized *java.JavaTypeParameterized, p any) java.JavaType
	VisitPrimitive(primitive *java.JavaTypePrimitive, p any) java.JavaType
	VisitUnknown(unknown *java.JavaTypeUnknown, p any) java.JavaType
	VisitVariable(variable *java.JavaTypeVariable, p any) java.JavaType
}

func (v *JavaTypeVisitor) self() JavaTypeVisitorI {
	if v.Self != nil {
		return v.Self
	}
	return v
}

// Visit dispatches to the appropriate visit method based on concrete type.
func (v *JavaTypeVisitor) Visit(javaType java.JavaType, p any) java.JavaType {
	if javaType == nil {
		return nil
	}

	switch t := javaType.(type) {
	case *java.JavaTypeAnnotation:
		return v.self().VisitAnnotation(t, p)
	case *java.JavaTypeArray:
		return v.self().VisitArray(t, p)
	case *java.JavaTypeShallowClass:
		return v.self().VisitShallowClass(t, p)
	case *java.JavaTypeClass:
		return v.self().VisitClass(t, p)
	case *java.JavaTypeGenericTypeVariable:
		return v.self().VisitGenericTypeVariable(t, p)
	case *java.JavaTypeIntersection:
		return v.self().VisitIntersection(t, p)
	case *java.JavaTypeMethod:
		return v.self().VisitMethod(t, p)
	case *java.JavaTypeMultiCatch:
		return v.self().VisitMultiCatch(t, p)
	case *java.JavaTypeParameterized:
		return v.self().VisitParameterized(t, p)
	case *java.JavaTypePrimitive:
		return v.self().VisitPrimitive(t, p)
	case *java.JavaTypeUnknown:
		return v.self().VisitUnknown(t, p)
	case *java.JavaTypeVariable:
		return v.self().VisitVariable(t, p)
	default:
		return javaType
	}
}

// VisitList visits a list of JavaTypes.
func (v *JavaTypeVisitor) VisitList(javaTypes []java.JavaType, p any) []java.JavaType {
	if javaTypes == nil {
		return nil
	}
	result := make([]java.JavaType, len(javaTypes))
	for i, jt := range javaTypes {
		result[i] = v.self().Visit(jt, p)
	}
	return result
}

// Default implementations — return the type unchanged.

func (v *JavaTypeVisitor) VisitAnnotation(annotation *java.JavaTypeAnnotation, p any) java.JavaType {
	return annotation
}

func (v *JavaTypeVisitor) VisitArray(array *java.JavaTypeArray, p any) java.JavaType {
	return array
}

func (v *JavaTypeVisitor) VisitClass(class_ *java.JavaTypeClass, p any) java.JavaType {
	return class_
}

func (v *JavaTypeVisitor) VisitShallowClass(shallow *java.JavaTypeShallowClass, p any) java.JavaType {
	return shallow
}

func (v *JavaTypeVisitor) VisitGenericTypeVariable(generic *java.JavaTypeGenericTypeVariable, p any) java.JavaType {
	return generic
}

func (v *JavaTypeVisitor) VisitIntersection(intersection *java.JavaTypeIntersection, p any) java.JavaType {
	return intersection
}

func (v *JavaTypeVisitor) VisitMethod(method *java.JavaTypeMethod, p any) java.JavaType {
	return method
}

func (v *JavaTypeVisitor) VisitMultiCatch(multiCatch *java.JavaTypeMultiCatch, p any) java.JavaType {
	return multiCatch
}

func (v *JavaTypeVisitor) VisitParameterized(parameterized *java.JavaTypeParameterized, p any) java.JavaType {
	return parameterized
}

func (v *JavaTypeVisitor) VisitPrimitive(primitive *java.JavaTypePrimitive, p any) java.JavaType {
	return primitive
}

func (v *JavaTypeVisitor) VisitUnknown(unknown *java.JavaTypeUnknown, p any) java.JavaType {
	return unknown
}

func (v *JavaTypeVisitor) VisitVariable(variable *java.JavaTypeVariable, p any) java.JavaType {
	return variable
}
